node/pkg/processor/reactor/reactor.go

```diff
@@ -0,0 +1,491 @@
+	c.logger = supervisor.Logger(ctx)
+	c.logger.With(zap.String("group", c.group))
+
+	c.stateMutex.Lock()
+	if c.state == StateFinalized {
+		c.stateMutex.Unlock()
+		return nil
+	}
+	c.stateMutex.Unlock()
+
+	tickFrequency := time.Second
+	ticker := time.NewTicker(tickFrequency)
+	defer ticker.Stop()
+
+	supervisor.Signal(ctx, supervisor.SignalHealthy)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case o := <-c.observationChannel:
+			c.observed(ctx, o)
+		case o := <-c.foreignObservationChannel:
+			c.observationReceived(ctx, o)
+		case <-ticker.C:
+			terminate, err := c.housekeeping(ctx)
+			if err != nil {
+				c.logger.Warn("housekeeping failed", zap.Error(err))
+			}
+			if terminate {
+				c.logger.Debug("reactor concluded; shutting down processing loop")
+				// Signal done such that the routine does not restart
+				// TODO this might leave garbage in the supervisor tree
+				supervisor.Signal(ctx, supervisor.SignalDone)
+				return nil
+			}
+		}
+	}
+}
+
+func (c *ConsensusReactor[K]) ObservationChannel() chan<- K {
+	return c.observationChannel
+}
+
+func (c *ConsensusReactor[K]) ForeignObservationChannel() chan<- *gossipv1.SignedObservation {
+	return c.foreignObservationChannel
+}
+
+// State returns the current state of the reactor
+func (c *ConsensusReactor[K]) State() State {
+	c.stateMutex.Lock()
+	defer c.stateMutex.Unlock()
+	return c.state
+}
+
+// VAASignatures returns the stored signatures in the order required by a VAA.
+func (c *ConsensusReactor[K]) VAASignatures() []*vaa.Signature {
+	c.stateMutex.Lock()
+	defer c.stateMutex.Unlock()
+	agg := make([]bool, len(c.gs.Keys))
+	var sigs []*vaa.Signature
+	for i, a := range c.gs.Keys {
+		s, ok := c.signatures[a]
+
+		if ok {
+			var bs [65]byte
+			if n := copy(bs[:], s); n != 65 {
+				panic(fmt.Sprintf("invalid sig len: %d", n))
+			}
+
+			sigs = append(sigs, &vaa.Signature{
+				Index:     uint8(i),
+				Signature: bs,
+			})
+		}
+
+		agg[i] = ok
+	}
+	return sigs
+}
+
+func (c *ConsensusReactor[K]) housekeeping(ctx context.Context) (bool, error) {
+	c.stateMutex.Lock()
+	defer c.stateMutex.Unlock()
+	defer c.logger.Debug("housekeeping completed")
+
+	c.logger.Debug("starting housekeeping")
+
+	switch c.state {
+	case StateUnobserved:
+		if time.Since(c.firstSeen) > c.config.UnobservedTimeout {
+			c.logger.Debug("timing out", zap.String("reason", "unobserved_timeout"))
+			// Time out
+			c.timeOut()
+		}
+	case StateObserved:
+		if time.Since(c.lastObservation) > c.config.QuorumTimeout {
+			c.logger.Debug("timing out", zap.String("reason", "quorum_timeout"))
+			// Time out
+			c.timeOut()
+		}
+
+		if time.Since(c.lastTransmission) > c.config.RetransmitFrequency {
+			// TODO backoff when transmission fails
+			c.logger.Debug("retransmitting", zap.String("reason", "unobserved_timeout"))
+			reactorResubmission.WithLabelValues(c.group).Inc()
+			// Retransmit signature
+			err := c.transmitSignature(ctx)
```

Comment:
Reobservation requests should happen on the layer above I think (i.e. the VAA processor) as they are highly payload specific.
I agree that retransmissions are only slightly useful if the transport layer is unreliable and a previous observation was missed.

Reply:
I agree that putting them in a higher layer makes sense.  I'm fine with doing this in a follow-up change but if we end up cutting a release that doesn't have it we should let the community managers know ahead of time so that they can prepare for doing manual re-observations.

Reply:
Added them in the processor